lib/src/util/unrelated_types_visitor.dart

```diff
@@ -4,119 +4,131 @@
-  }
-
-  return _findIterableTypeArgument(definition, type.superclass,
-      accumulator: [type, ...accumulator, ...implementedInterfaces]);
-}
-
-bool _isParameterizedMethodInvocation(
-        String methodName, MethodInvocation node) =>
-    node.methodName.name == methodName &&
-    node.argumentList.arguments.length == 1;
-
 /// Base class for visitor used in rules where we want to lint about invoking
 /// methods on generic classes where the type of the singular argument is
 /// unrelated to the singular type argument of the class. Extending this
-/// visitor is as simple as knowing the method, class and library that uniquely
-/// define the target, i.e. implement only [definition] and [methodName].
+/// visitor is as simple as knowing the methods, classes and libraries that
+/// uniquely define the target, i.e. implement only [methods].
 abstract class UnrelatedTypesProcessors extends SimpleAstVisitor<void> {
   final LintRule rule;
   final TypeSystem typeSystem;
+  final TypeProvider typeProvider;
 
-  UnrelatedTypesProcessors(this.rule, this.typeSystem);
-
-  /// The type definition which this [UnrelatedTypesProcessors] is concerned
-  /// with.
-  InterfaceTypeDefinition get definition;
+  UnrelatedTypesProcessors(this.rule, this.typeSystem, this.typeProvider);
 
-  /// The name of the method which this [UnrelatedTypesProcessors] is concerned
+  /// The method definitions which this [UnrelatedTypesProcessors] is concerned
   /// with.
-  String get methodName;
+  List<MethodDefinition> get methods;
 
   @override
   void visitMethodInvocation(MethodInvocation node) {
-    if (!_isParameterizedMethodInvocation(methodName, node)) {
+    if (node.argumentList.arguments.length != 1) {
+      return;
+    }
+    var methodDefinition = _matchingMethod(node);
+    if (methodDefinition == null) {
       return;
     }
 
     // At this point, we know that [node] is an invocation of a method which
-    // has the same name as the method that this UnrelatedTypesProcessors] is
-    // concerned with, and that the method has a single parameter.
+    // has the same name as the method that this [UnrelatedTypesProcessors] is
+    // concerned with, and that the method call has a single argument.
     //
     // We've completed the "cheap" checks, and must now continue with the
     // arduous task of determining whether the method target implements
     // [definition].
 
     DartType? targetType;
+    // TODO(srawlins): Should be `realTarget`?
     var target = node.target;
     if (target != null) {
       targetType = target.staticType;
     } else {
       for (AstNode? parent = node; parent != null; parent = parent.parent) {
         if (parent is ClassDeclaration) {
           targetType = parent.declaredElement2?.thisType;
+          break;
         } else if (parent is MixinDeclaration) {
           targetType = parent.declaredElement2?.thisType;
+          break;
         }
+        // TODO(srawlins): Extension? Enum?
```

Comment:
Extensions don't define a type, but enums do.

Reply:
But the receiver in an extension can still be implicit.

Reply:
Ok, I understand this code better now. Yes, for extension types you'll want to use the `on` type.

Reply:
Done in other PRs